dissect/target/tools/query.py

```diff
@@ -110,67 +118,127 @@ def main():
                 "The --rewrite-cache option will be ignored as --no-cache or --only-read-cache are specified",
             )
 
+    # -l/-f follow the same code path, a list = a list of functions + a display command
+    # this makes sure the behavior will always be in sync
+    if args.list is not False:
+        args.function = args.list
+
+    # If no functions (or list) then we cannot continue at all
+    if not args.function:
+        parser.error("argument -f/--function is required")
+
+    # No targets is allowed, but only for lists/help
+    if not args.targets and args.list is False and "-h" not in rest and "--help" not in rest:
+        parser.error("too few arguments")
+
+    # In that case the user wants to see all available plugins and loaders
+    # and we turn off compatibility checks, because they will always fail
+    # we don't raise an exception though, because the user might want to 'amend'
+    # the command later...
+    if not args.targets:
+        targets = [Target()]
+        args.compatible = False
+    else:
+        targets = Target.open_all(args.targets, args.children)
+
+    # To check whether the output is uniform and which plugins we need to load
+    # either for the list or for execution, we collect all the work to be done here
+    tasklist = []
+    for target in targets:
+        if args.child:
+            try:
+                target = target.open_child(args.child)
+            except Exception:
+                continue
+        functions = target.find_plugin_functions(args.function, args.compatible)
+        tasklist.append((target, functions))
+
+    # Show the list of available plugins for the given optional target and optional
+    # search pattern, only display plugins that can be applied to ALL targets
     if args.list:
+        collected_plugins = None
+        for index, task in enumerate(tasklist):
+            collected = {}
+            target, funcs = task
+            for func in funcs:
+                collected[func.name] = func.plugin_desc
+
+            if collected_plugins:
+                common = collected_plugins.keys() & collected.keys()
+                collected_plugins = {plugin_name: collected_plugins[plugin_name] for plugin_name in common}
+            else:
+                collected_plugins = collected
+
+        # Display in a user friendly manner
         t = Target()
         fparser = generate_argparse_for_bound_method(t.plugins, usage_tmpl=USAGE_FORMAT_TMPL)
         fargs, rest = fparser.parse_known_args(rest)
-        t.plugins(**vars(fargs))
-
-        fparser = generate_argparse_for_bound_method(t.loaders, usage_tmpl=USAGE_FORMAT_TMPL)
-        fargs, rest = fparser.parse_known_args(rest)
-        t.loaders(**vars(fargs))
-
+        t.plugins(collected_plugins.values())  # , True, **vars(fargs)
+        # No real targets specified, user might be interested to see available loaders
+        if len(tasklist) == 1 and not tasklist[0][0]._applied:
```

Comment:
Does this matter? You use the fake `t` Target here anyway.

Reply:
That is not the issue, I thought it would be nice to only show the loaders if there are no targets. I refactored this code because it was ugly btw. Loaders take up a lot of space in my terminal, should we always display that chunk?